### ISSUE TITLE
Triage: detecting missing or unknown commands

### ISSUE THREAD

**Issue by shadowspawn** [tj/commander.js#1088]
What is the best way to test for missing or unknown commands? Is the existing support good enough?

Research prompted by Commander v4 which changes `program.args` and will break some client code which was dependent on the previous behaviour: #1048

Example of affected code: https://github.com/tj/commander.js/issues/7#issuecomment-32448653

Related issues:
- #7 Show help if no command was executed
- #240 Support to know if any known command was processed, or any output written
- #338 How to detect an unsupported command? [git-style]
- #432 running an unknown subcommand does not fail
- #603 Show help when command is unknown
- #809 Using sub-command always invokes *
- #940 argments are empty
- #942 Throw error when no sub-command specified
- #977 It looks not way to know has action is run?
- #978 Top level command action does not work
- #1032 Passing command option leaves content in args

**Comment by shadowspawn** [comment]
A direct test for no arguments at all (including options) is used in the README. This works the same in Commander v4.
```
if (!process.argv.slice(2).length) {
  program.outputHelp(make_red);
}
```

or without the slice:
```
if (process.argv.length < 3) {
  program.help();
}
```

or equivalently using `rawArgs:
```
if (program.rawArgs.length < 3) {
  program.help();
}
```

**Comment by shadowspawn** [comment]
The example code from the README for detected unknown commands (in a program using action handlers) works the same in Commander v4:

```
program.on('command:*', function () {
  console.error('Invalid command: %s\nSee --help for a list of available commands.', program.args.join(' '));
  process.exit(1);
});
```

**Comment by shadowspawn** [comment]
I am somewhat interested in adding support for testing for an unhandled command, which has not triggered any action handlers or external commands. This would implicitly cover both missing and unknown commands with a single test.

**Comment by shadowspawn** [comment]
People are also interested in having Commander do the messaging in missing and unknown cases, but probably treat that as a separate enhancement.

**Comment by shadowspawn** [comment]
By default, an unknown subcommand shows an error and a missing subcommand shows the help in the published pre-release of Commander v5.0.0 (#1163)
